### PR TITLE
🗻 autograd: server-side gradient processing 

### PR DESCRIPTION
**Pull request by tylerflex**
currently a WIP, needs backend bits and pieces.

For now, organizes the autograd pipeline into a local and a remote pathway, depending on a kwarg passed to `run()`.

The local one operates the same as before, the remote one has the basics of what we need to implement server side. Specifically, I would need the following Web API implemented.

```py
sim_data = web.run(sim_combined, simulation_type="autograd_fwd")
```
on the `sim_combined` (sim with original + grad monitors) should split the data into `original` and `fwd` and store the `fwd` associated with this task ID. Then return just the original `SimulationData`.

```py
sim_data = web.run(sim_adj, simulation_type="autograd_bwd", parent_tasks=[task_id_fwd])
```

should run the adjoint simulation `sim_adj` and compute the gradient information combined with the cached forward `SimulationData` associated with `task_id_fwd `

```py
fields_vjp = download_vjp_fields_autograd(task_id_adj)
```

should download the result of the adjoint post processing associated with the task ID from the `autograd_bwd` call above.

I will work to make exactly what Neds to happen in each of these tasks explicit. it's just reorganizing bits of the existing front end code.

### CODE REVIEW

File: tidy3d/web/api/autograd/autograd.py [hunk 1]
```diff
@@ -27,6 +29,9 @@
 
 MAX_NUM_TRACED_STRUCTURES = 500
 
+# default value for whether to do local gradient calculation (True) or server side (False)
+LOCAL_GRADIENT = True
```
Comment (momchil-flex) [inline tidy3d/web/api/autograd/autograd.py:37]: We will set it to False eventually?

File: tests/test_components/test_autograd.py [hunk 2]
```diff
@@ -114,15 +115,84 @@ def use_emulated_run(monkeypatch):
...
+
+                aux_data = {}
+                _ = postprocess_fwd(
+                    sim_data_combined=sim_data_combined,
+                    sim_original=sim_original,
+                    aux_data=aux_data,
+                )
+
+                sim_data_orig = aux_data[AUX_KEY_SIM_DATA_ORIGINAL]
+                sim_data_fwd = aux_data[AUX_KEY_SIM_DATA_FWD]
```
Comment (momchil-flex) [inline tests/test_components/test_autograd.py]: In adjoint, there's a call for this like

```
        sim_data, sim_data_store = JaxSimulationData.split_fwd_sim_data(
            sim_data=sim_data, jax_info=jax_info
        )
```

**Review by momchil-flex — COMMENTED** [review]
[This](https://github.com/flexcompute/tidy3d-core/blob/75631b7cfea0e0451920fb804d5f2478af2f98cd/tidy3d_backend/run.py#L340-L363) is the gist of the adjoint handling on the backend, apart from some things like ensuring the right files are downloaded which happen elsewhere. It's a bit hard for me to orient myself here yet but in the end it should reduce to something similar.

File: tidy3d/web/api/autograd/autograd.py (hunk 1 above)
Comment (tylerflex) [reply tidy3d/web/api/autograd/autograd.py:37]: Yea, setting to True until this works :)

File: tests/test_components/test_autograd.py (hunk 2 above)
Comment (tylerflex) [reply tests/test_components/test_autograd.py]: yea, I guess I could refactor this into it's own method of `Simulation` if it would make things easier.

**Comment by tylerflex** [comment]
> [This](https://github.com/flexcompute/tidy3d-core/blob/75631b7cfea0e0451920fb804d5f2478af2f98cd/tidy3d_backend/run.py#L340-L363) is the gist of the adjoint handling on the backend, apart from some things like ensuring the right files are downloaded which happen elsewhere. It's a bit hard for me to orient myself here yet but in the end it should reduce to something similar.

yea I tried to emulate that a bit in this part of the emulated `Job.run()`

https://github.com/flexcompute/tidy3d/blob/81c554465c4a32b1d625b865ecbb1d51e3838911/tests/test_components/test_autograd.py#L145-L187

**Comment by tylerflex** [comment]
@momchil-flex I want to make sure this doesn't fall through the cracks for 2.8 since probably it will be highly requested.. do you think you'll find time to work on the backend bits? is there anything I can do to make it easier?

**Comment by momchil-flex** [comment]
> @momchil-flex I want to make sure this doesn't fall through the cracks for 2.8 since probably it will be highly requested.. do you think you'll find time to work on the backend bits? is there anything I can do to make it easier?

I'll give it a try tomorrow. Meanwhile do you want to try to introduce the `split_fwd_sim_data` method from the comment above and remove the "bit of a hack" comment in the code about this?

**Comment by tylerflex** [comment]
> Meanwhile do you want to try to introduce the split_fwd_sim_data method from the comment above and remove the "bit of a hack" comment in the code about this?

Thanks. I won't be able to get to this until tomorrow eastern US time. I guess just let me know if you've already started so we don't conflict.

**Comment by tylerflex** [comment]
@momchil-flex I pushed this [commit](https://github.com/flexcompute/tidy3d/pull/1760/commits/ca401adebfa3b381d10c32d25eb3e6921d0d538e) that reorganizes things a bit.

Adds:
```py
sim_data_original, sim_data_fwd = sim_data_combined.split_original_fwd(num_mnts_original=num_mnts_original)
```
although I'm not sure you'll need this directly as it's done in `postprocess_fwd` ([code](https://github.com/flexcompute/tidy3d/blob/tyler/autograd_/serverside/tidy3d/web/api/autograd/autograd.py#L466-L487)).

Removes the "kind of a hack" comment, basically needed a way to figure out what were original vs gradient monitors. But I restructured stuff so now one has access to the original sim directly.

A few other small improvements and cleanup.

Should be set. Note that the async bits are not done in this yet, maybe we can start with regular `web.run()`

**Comment by tylerflex** [comment]
@momchil-flex I reworked this according to our comments and also rebased against develop. could you revisit when you get a chance? thanks

**Comment by tylerflex** [comment]
done in #1824

**Pull request closed without merging**